src/main/java/practice/StreamPractice.java

```diff
@@ -1,80 +1,115 @@
-     * have nationality - 'Ukrainian'
-     * and live in Ukraine for 10 years. For the last requirement use field periodsInUkr,
-     * which has following view: "2002-2015"
-     * We want to reuse our validation in future, so let's write our own impl of Predicate
-     * parametrized with Candidate in CandidateValidator.
-     */
-    public List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
-    }
-}
+package practice;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Predicate;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import model.Candidate;
+import model.Cat;
+import model.Person;
+
+public class StreamPractice {
+    /**
+     * Given list of strings where each element contains 1+ numbers:
+     * input = {"5,30,100", "0,22,7", ...}
+     * return min integer value. One more thing - we're interested in even numbers.
+     * If there is no needed data throw RuntimeException with message
+     * "Can't get min value from list: < Here is our input 'numbers' >"
+     */
+    public int findMinEvenNumber(List<String> numbers) {
+        return numbers.stream()
+                .map(number -> number.split(","))
+                .flatMap(Stream::of)
+                .mapToInt(Integer::parseInt)
+                .filter(num -> num % 2 == 0)
+                .min()
+                .orElseThrow(() -> new RuntimeException("Can't get min value from list: "
+                        + numbers));
+    }
+
+    /**
+     * Given a List of Integer numbers,
+     * return the average of all odd numbers from the list or throw NoSuchElementException.
+     * But before that subtract 1 from each element on an odd position (having the odd index).
+     */
+    public Double getOddNumsAverage(List<Integer> numbers) {
+        return IntStream.range(0, numbers.size())
+                        .map(index -> index % 2 != 0 ? numbers.get(index) - 1 : numbers.get(index))
+                        .filter(number -> number % 2 != 0)
+                        .average()
+                        .orElseThrow(NoSuchElementException::new);
+    }
+
+    /**
+     * Given a List of `Person` instances (having `name`, `age` and `sex` fields),
+     * for example, `Arrays.asList( new Person(«Victor», 16, Sex.MAN),
+     * new Person(«Helen», 42, Sex.WOMAN))`,
+     * select from the List only men whose age is from `fromAge` to `toAge` inclusively.
+     * <p>
+     * Example: select men who can be recruited to army (from 18 to 27 years old inclusively).
+     */
+    public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toAge) {
+        return peopleList.stream()
+                .filter(people -> people.getAge() >= fromAge && people.getAge() <= toAge
+                        && people.getSex() == Person.Sex.MAN)

```

Comment:
a good approach is to give lambda parameters as short names as possible
```suggestion
                .filter(p-> p.getAge() >= fromAge && p.getAge() <= toAge && p.getSex() == Person.Sex.MAN)

```